### ISSUE TITLE
Shortcut for making relative URLs absolute.

### ISSUE THREAD

**Issue by EliotJones** [AngleSharp/AngleSharp#441]
When cropping part of an HTML document into a new document created as follows:

```
var target = await (IHtmlDocument)BrowsingContext.New().OpenAsync(x =>
{
    x.Content("<!doctype html>");
    x.Address(originalUrl);
}).
```

I want to change all relative image and CSS URLs to absolute URLs in the target document. I can add the copied region with:

```
target.Body.Append(croppedRegion.Body.ChildNodes.ToArray());
```

Then loop through all images, links etc in the following way:

```
var images = target.Images;

foreach (var image in images)
{
    image.Source = image.Source;
}
```

This will set the relative URL to an absolute URL but it seems like a nasty way of achieving the result. Is there a better way?

**Comment by FlorianRappl** [comment]
My recommendation would be to use `image.SetAttribute("src", image.Source)` instead, but its still some nasty stuff - right.

There is no other way - if I would provide one it would be in form of an extension method. However, I think you would be better off using your own solution, as you may (or may not) want to include certain elements.

I will leave this open (I think it would be an interesting extension method and maybe I'll propose a signature / behavior here).

**Comment by EliotJones** [comment]
Thanks for your response. I think your suggested form is more descriptive / less confusing for future maintainers, but as I have tests around this part and commented the usage it should be understandable to anyone who comes across it.

It's probably not a common use case and there's a decent workaround so feel free to close the issue but like you say it's an interesting design challenge to work out the best API for it. My initial thought is it could be some option on the `document.ToString()` (or HtmlDocument specific ToString() style method I have forgotten the name of) but this would restrict it to use cases where the document is being output as a string.

**Comment by FlorianRappl** [comment]
I'll leave it open for now. I am not sure if / when I have a good (generic) solution for this.

**Comment by FlorianRappl** [comment]
After giving it much thought today I don't think there can / will be a generic version for this. There are many reasons:

- I can add a helper to do the transformation, but this would be a one-time thing (i.e., if something changes then the new URLs may be relative again; not very good experience)
- What is an URL is very domain specific, i.e., you care about the image URLs, but others may care about the general hyperlinks, and somebody else may care only about URLs referencing stylesheets, ... going over all URLs later will suffer from having to know *all* places where a URL is used (a better way is to embed this in the URL representation algorithm, however, then we need an option for this and perf. will suffer)

Long story short I think that this feature would be too domain specific and cannot be easily generalized. I'll close the issue for now.
